Rewrite SIMD article as "SIMD and Hardware Intrinsics" (#54834)
dotnet/docs: docs/standard/simd.md

---
title: Use SIMD and Hardware Intrinsics in .NET
description: Learn about the layered SIMD support in .NET, from System.Numerics vector types to cross-platform Vector64/128/256/512 APIs, hardware intrinsics, and TensorPrimitives.
author: FIVIL
ms.author: tagoo
ms.date: 07/18/2026
ai-usage: ai-assisted
---

> # Use SIMD and Hardware Intrinsics in .NET

Reviewer: ```suggestion
# Use SIMD and hardware intrinsics in .NET
```

.NET provides several kinds of SIMD support. Pick the one that matches how much control you need and how much complexity you're willing to take on.

## Types of SIMD support in .NET

| API | Namespace | When to use it |
| --- | --- | --- |
| Fixed-purpose vector and matrix types | <xref:System.Numerics?displayProperty=fullName> | Graphics and geometry math with 2-4 element vectors, matrices, quaternions, and planes. |
| `Vector<T>` | <xref:System.Numerics?displayProperty=fullName> | Portable, variable-width vectorization when you don't need per-platform control. |
| `Vector64<T>`, `Vector128<T>`, `Vector256<T>`, `Vector512<T>` | <xref:System.Runtime.Intrinsics?displayProperty=fullName> | Cross-platform, fixed-width vectorization with fine-grained control. This is the recommended starting point for new vectorized algorithms. |
| Hardware intrinsics | <xref:System.Runtime.Intrinsics.X86?displayProperty=fullName>, <xref:System.Runtime.Intrinsics.Arm?displayProperty=fullName>, <xref:System.Runtime.Intrinsics.Wasm?displayProperty=fullName> | Specific processor instructions that the higher-level APIs don't expose, for the last bit of performance on a hot path. |

> | `TensorPrimitives` | <xref:System.Numerics.Tensors?displayProperty=fullName> | Ready-made, vectorized math over spans. It does the vectorization for you. |

Reviewer: ```suggestion
| Fixed-purpose vector and matrix types | <xref:System.Numerics> | Graphics and geometry math with 2-4 element vectors, matrices, quaternions, and planes. |
| `Vector<T>` | <xref:System.Numerics> | Portable, variable-width vectorization when you don't need per-platform control. |
| `Vector64<T>`, `Vector128<T>`, `Vector256<T>`, `Vector512<T>` | <xref:System.Runtime.Intrinsics> | Cross-platform, fixed-width vectorization with fine-grained control. This is the recommended starting point for new vectorized algorithms. |
| Hardware intrinsics | <xref:System.Runtime.Intrinsics.X86>, <xref:System.Runtime.Intrinsics.Arm?displayProperty=fullName>, <xref:System.Runtime.Intrinsics.Wasm?displayProperty=fullName> | Specific processor instructions that the higher-level APIs don't expose, for the last bit of performance on a hot path. |
| `TensorPrimitives` | <xref:System.Numerics.Tensors> | Ready-made, vectorized math over spans. It does the vectorization for you. |
```

Namespaces are always displayed fully qualified.

## Check for hardware acceleration

SIMD-accelerated types work even on hardware or JIT configurations that don't support SIMD, because they fall back to non-accelerated software implementations. To branch on whether acceleration is actually available, check the relevant `IsHardwareAccelerated` property:

- <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType> reports whether `Vector<T>` operations are accelerated.
- <xref:System.Runtime.Intrinsics.Vector128.IsHardwareAccelerated?displayProperty=nameWithType>, and the `Vector64`/`Vector256`/`Vector512` equivalents, report acceleration for each fixed width.

These properties are turned into constants by the JIT, so the branches you don't take are eliminated and there's no runtime cost to checking them. Don't cache the values; read them directly where you need them. The same applies to the `Count` properties (for example, `Vector128<T>.Count`), which are also JIT-time constants.

> Most operations on an accelerated width are themselves accelerated, but it isn't guaranteed for every operation. For example, floating-point division may be accelerated where integer division isn't. When `Vector256` is accelerated, `Vector128` usually is too, but there's no guarantee, so check each width you use.

Reviewer: ```suggestion
Most operations on an accelerated width are themselves accelerated, but it isn't guaranteed for every operation. For example, floating-point division might be accelerated where integer division isn't. When `Vector256` is accelerated, `Vector128` usually is too, but there's no guarantee, so check each width you use.
```

> [!TIP]
> If an operation you need isn't accelerated on a platform you care about, or you'd like a new cross-platform API, file an issue on [dotnet/runtime](https://github.com/dotnet/runtime/issues). The same applies to codegen improvements.

> Not every element type is valid for every vector. `Vector128<T>` and its siblings support the primitive numeric types (`byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `nint`, and `nuint`) today, and that set may grow to include other types in the future. Use `Vector128<T>.IsSupported` to determine whether a given `T` is valid, which is especially useful from generic code.

Reviewer: ```suggestion
Not every element type is valid for every vector. `Vector128<T>` and its siblings support the primitive numeric types (`byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `nint`, and `nuint`) today, and that set might grow to include other types in the future. Use `Vector128<T>.IsSupported` to determine whether a given `T` is valid, which is especially useful from generic code.
```

## Cross-platform vectorization with Vector128

`Vector128<T>` is the common denominator across every platform that supports vectorization, so it's the best place to start. It holds a 128-bit vector: 16 bytes, 8 shorts, 4 ints/floats, or 2 longs/doubles. `Vector256<T>` is twice as wide, and `Vector512<T>` twice again. Not all hardware supports the larger widths, so the examples that follow use `Vector128` for portability.

Each width has a generic type (`Vector128<T>`) for the data and a non-generic static class (<xref:System.Runtime.Intrinsics.Vector128>) that holds most of the operations, including static factory methods like `Create` and `Load`. Operators such as `+`, `&`, and `<<` are the idiomatic way to express arithmetic and bit operations; prefer them over the named method equivalents to avoid operator-precedence bugs and improve readability. For algorithms that depend on byte order, branch on <xref:System.BitConverter.IsLittleEndian>, which the JIT also folds to a constant.

> [!NOTE]

> > On x86/x64, `Vector256<T>` operations are generally treated as two independent 128-bit "lanes". For most element-wise operations this is transparent, but operations that cross lanes (such as shuffles or pairwise/horizontal operations) may behave differently or cost more than the `Vector128` equivalent. Confirm with benchmarks before assuming a wider vector is faster.

Reviewer: ```suggestion
> On x86/x64, `Vector256<T>` operations are generally treated as two independent 128-bit "lanes". For most element-wise operations this is transparent, but operations that cross lanes (such as shuffles or pairwise/horizontal operations) might behave differently or cost more than the `Vector128` equivalent. Confirm with benchmarks before assuming a wider vector is faster.
```

### Common operations

> `Vector128` and its wider siblings expose a large API surface. You don't need to memorize it—know the categories and look up the details when you need them. Every operation has a software fallback for platforms that can't accelerate it. The table below covers essentially the whole surface.

Reviewer: ```suggestion
`Vector128` and its wider siblings expose a large API surface. You don't need to memorize it—know the categories and look up the details when you need them. Every operation has a software fallback for platforms that can't accelerate it. The following table covers essentially the whole surface.
```

| System.Numerics interop | Reinterpret between `Vector128<T>` and the fixed-shape numerics types | `AsVector`, `AsVector2`, `AsVector3`, `AsVector4`, `AsPlane`, `AsQuaternion`, `AsVector128`, `AsVector128Unsafe` |
| Reorder | Rearrange elements by index or interleave two vectors | `Shuffle`, `Reverse`, `Zip`, `ZipLower`, `ZipUpper`, `Unzip`, `UnzipEven`, `UnzipOdd`, `ConcatLowerLower`, `ConcatLowerUpper`, `ConcatUpperLower`, `ConcatUpperUpper` |
| Lane access | Read or replace individual elements and halves, or resize the vector | `GetElement`, `WithElement`, `ToScalar`, `GetLower`, `GetUpper`, `WithLower`, `WithUpper`, `ToVector256` |

> [!TIP]
> Several operations have `Estimate` and `Native` variants—for example, `MultiplyAddEstimate`, `ClampNative`, `MinNative`, `MaxNative`, `ShuffleNative`, and `ConvertToInt32Native`. They map to a faster hardware instruction that trades some precision or gives up an IEEE edge-case guarantee (such as NaN handling), so reach for them only when a benchmark shows the exact form is the bottleneck and the looser semantics are acceptable.

> [!NOTE]

> > `Vector256.Shuffle` treats its input as a single 256-bit vector, where-as the platform-specific `Avx2.Shuffle` operates as two independent 128-bit lanes. The cross-platform API is the more portable choice, but confirm the behavior you need when porting hand-written intrinsics.

Author: I can remove this if required, but it is rather a personal writing quirk of myself (the author) and such quirks are prevalent in books and text throughout history.

Trying to normalize everything is not strictly a positive, IMO.

:::code language="csharp" source="./snippets/simd/csharp/AdvancedVectorization.cs" id="CodeStructure":::

Each width's outer guard combines `Vector128.IsHardwareAccelerated` (a JIT-time constant for whether the platform accelerates that width) with `Vector128<T>.IsSupported` (whether the element type `T` is valid for that width). Inside a supported block, compare the input length against `Count` to choose between the vectorized path and a small-input fallback. The method is generic over `T`, and the `Vector256` and `Vector512` blocks—identical to the `Vector128` block but using the wider type—are shown commented out for brevity.

There are two distinct fallbacks. A buffer that's too small for even the narrowest vector, but on accelerated hardware, goes to `SumVectorSmall`—an explicit `switch` jump table that handles each possible sub-vector length without a loop:

:::code language="csharp" source="./snippets/simd/csharp/AdvancedVectorization.cs" id="VectorSmall":::

> `sizeof(T)` is also a JIT-time constant (and, with the `unmanaged` constraint, doesn't require an `unsafe` context in C# 15 / .NET 11), so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. Only the 4-byte table is shown; the 1-byte, 2-byte, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.

Reviewer 2: We (ECMA committee) introduced a spec bug when we added the `unmanaged` constraint that we recently fixed. See this PR: https://github.com/dotnet/csharpstandard/pull/1721.

The new relaxed rules for `sizeof(T)` are part of the new unsafe model. That should still need the `preview` language  version in C# 15 / .NET 11. Otherwise, only the named built in types are allowed for sizeof in a safe context. We should call that out in the first sentence. My suggestion splits it into 2 sentences because it gets harder to read:

```suggestion
`sizeof(T)` is also a JIT-time constant, so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. (When the new memory safety model is enabled, `sizeof(T)` expressions on a type parameter with the `unmanaged` constraint is allowed in safe code.) Only the 4-byte table is shown; the 1-byte, 2-byte, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.
```

Author: @BillWagner aren't the `sizeof(T)` relaxations part of the general change to make pointers no longer unsafe and are not part of the unsafe-v2 model that is in preview and requires opt-in?

Reviewer 2: I checked with @333fred.  These should only light up with LangVersion.= preview, even after .NET 11 G.A.

> `sizeof(T)` is also a JIT-time constant (and, with the `unmanaged` constraint, doesn't require an `unsafe` context in C# 15 / .NET 11), so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. Only the 4-byte table is shown; the 1-, 2-, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.

Reviewer: ```suggestion
`sizeof(T)` is also a JIT-time constant (and, with the `unmanaged` constraint, doesn't require an `unsafe` context in C# 15 / .NET 11), so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. Only the 4-byte table is shown; the 1-byte, 2-byte, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.
```

Reviewer: 1-byte, 2-byte, and 8-byte

An **idempotent** operation such as searching for a value can reprocess the overlap harmlessly, so it folds the last vector in directly with no mask:

:::code language="csharp" source="./snippets/simd/csharp/SimpleVectorization.cs" id="VectorizedRemainder":::

> [!WARNING]
> Mishandling the remainder is a common source of bugs. A loop that reads past the end of the buffer produces non-deterministic results and can crash. The runtime's test suite uses a `BoundedMemory` helper that places a no-access page immediately after the buffer, so any out-of-bounds read throws an <xref:System.AccessViolationException> during testing. Always cover the remainder logic, including buffers whose length isn't a multiple of the vector width.

### Load and store vectors safely

> For most code, `Vector128.Create(span)` and <xref:System.Runtime.Intrinsics.Vector128.CopyTo*> are the simplest way to move data between a span and a vector, and the JIT keeps them efficient. When you need the lower-level load and store—for example, to walk a buffer by managed reference—prefer the <xref:System.Runtime.Intrinsics.Vector128.LoadUnsafe*> and <xref:System.Runtime.Intrinsics.Vector128.StoreUnsafe*> overloads that take a managed reference and an `nuint` element offset. Unlike the pointer-based `Load`/`Store` overloads, they don't require pinning the buffer, and unlike raw reference arithmetic they don't need you to manually advance a `ref`. Both alternatives are easy to get wrong in ways that introduce garbage-collector holes or access violations.

Reviewer: ```suggestion
For most code, `Vector128.Create(span)` and <xref:System.Runtime.Intrinsics.Vector128.CopyTo*> are the simplest way to move data between a span and a vector, and the JIT keeps them efficient. When you need the lower-level load and store—for example, to walk a buffer by managed reference—prefer the <xref:System.Runtime.Intrinsics.Vector128.LoadUnsafe*> and <xref:System.Runtime.Intrinsics.Vector128.StoreUnsafe*> overloads that take a managed reference and an `nuint` element offset. Unlike the pointer-based `Load`/`Store` overloads, they don't require pinning the buffer, and unlike raw reference arithmetic, they don't need you to manually advance a `ref`. Both alternatives are easy to get wrong in ways that introduce garbage-collector holes or access violations.
```

To exercise every path on a single machine, run the test suite once with no overrides, once with `DOTNET_EnableAVX2=0`, and once with `DOTNET_EnableHWIntrinsic=0`. The alternative is running across enough varied hardware to cover them.

### Instruction-set configuration knobs

Beyond those two, the runtime recognizes a knob per logical grouping of instruction sets, each prefixed with `DOTNET_`. A single knob can cover several related instruction sets—`EnableAVX2`, for instance, gates AVX2 along with BMI1, BMI2, F16C, FMA, LZCNT, and MOVBE. Setting a knob to `0` disables its whole group and everything layered on top of it. Setting it to `1` (the default for most) allows the group, but the hardware must still actually support it—enabling a knob the current CPU lacks is ignored, so you can only ever narrow what's used, never force an unsupported instruction on and break yourself. `DOTNET_EnableHWIntrinsic=0` is the big hammer—it turns off everything down to the base, so `Vector128`, `Vector64`, and `Vector<T>` all report no acceleration and the code falls to its software path.

> [!IMPORTANT]

> > These are diagnostic tools, meant primarily for testing and validation—exercising each code path, reproducing a hardware-specific issue, or confirming a fallback. They aren't designed for general or production use, and they aren't a stability contract: the set below is what .NET 11 recognizes; earlier releases exposed a different set—the baseline and AVX-512 knobs in particular were reconfigured—so confirm the names against the runtime version you target.

Reviewer: ```suggestion
> These are diagnostic tools, meant primarily for testing and validation—exercising each code path, reproducing a hardware-specific issue, or confirming a fallback. They aren't designed for general or production use, and they aren't a stability contract. The following set is what .NET 11 recognizes; earlier releases exposed a different set—the baseline and AVX-512 knobs in particular were reconfigured—so confirm the names against the runtime version you target.
```

>

> > They also have limits on what they reach. Because they gate JIT decisions, they don't affect code that was already compiled ahead of time through ReadyToRun or Native AOT, and they don't necessarily affect internal routines the runtime and core libraries use themselves. Treat them as a way to steer your own JIT-compiled code, not a global off switch for an instruction set.

Reviewer: ```suggestion
> These tools also have limits on what they reach. Because they gate JIT decisions, they don't affect code that was already compiled ahead of time through ReadyToRun or Native AOT, and they don't necessarily affect internal routines the runtime and core libraries use themselves. Treat them as a way to steer your own JIT-compiled code, not a global off switch for an instruction set.
```

Each logical grouping of x86/x64 instruction sets also has its own switch:

| Knob (`DOTNET_` prefix) | Default | Gates |
| --- | --- | --- |
| `EnableAVX` | `1` | AVX and dependents |
| `EnableAVX2` | `1` | AVX2, BMI1, BMI2, F16C, FMA, LZCNT, MOVBE, and dependents |
| `EnableAVX512` | `1` | AVX-512 F+BW+CD+DQ+VL and dependents |
| `EnableAVX512BMM` | `1` | AVX-512 BMM |
| `EnableAVX512v2` | `1` | AVX-512 IFMA+VBMI |
| `EnableAVX512v3` | `1` | AVX-512 BITALG+VBMI2+VPOPCNTDQ+VNNI |
| `EnableAVX10v1` | `1` | AVX10.1 |
| `EnableAVX10v2` | `0` | AVX10.2 |
| `EnableAPX` | `0` | APX (extended general-purpose registers) |
| `EnableAES` | `1` | AES, PCLMULQDQ |
| `EnableAVX512VP2INTERSECT` | `1` | AVX-512 VP2INTERSECT |
| `EnableAVXIFMA` | `1` | AVX-IFMA |
| `EnableAVXVNNI` | `1` | AVX-VNNI |
| `EnableAVXVNNIINT` | `1` | VEX AVX-VNNI-INT8 and AVX-VNNI-INT16 |
| `EnableGFNI` | `1` | GFNI |
| `EnableSHA` | `1` | SHA |
| `EnableVAES` | `1` | VAES, VPCLMULQDQ |

> | `EnableWAITPKG` | `1` | WAITPKG |

Reviewer 3: Some instruction sets like these, are not yet available as intrinsics.

Author: This is somewhat intentional. I'm 100% going to forget to update this to be in sync all the time, so I'd rather give out the full list including the ones we already have detection support for and will be lighting up in .NET 12.

| `EnableX86Serialize` | `1` | X86 SERIALIZE |

On Arm64, each logical grouping of instruction sets has its own switch:

| Knob (`DOTNET_` prefix) | Default | Gates |
| --- | --- | --- |
| `EnableArm64Aes` | `1` | AES |
| `EnableArm64Atomics` | `1` | Large System Extensions (LSE) atomics |
| `EnableArm64Crc32` | `1` | CRC32 |
| `EnableArm64Dczva` | `1` | `DC ZVA` cache-zeroing |
| `EnableArm64Dp` | `1` | Dot Product |

> | `EnableArm64Rdm` | `1` | Rounding Double Multiply (RDMA) |

Author: Fixed -- it's `RDM` (Rounding Doubling Multiply Accumulate); corrected to match the knob name.

> [!NOTE]
> Reply posted by Copilot on my behalf.

| `EnableArm64Sha1` | `1` | SHA1 |
| `EnableArm64Sha256` | `1` | SHA256 |

> | `EnableArm64Rcpc` | `1` | Release Consistent processor consistent (RCPC) |

Author: Fixed -- dropped the duplicated word and corrected the casing to `RCpc`/`RCpc2`.

> [!NOTE]
> Reply posted by Copilot on my behalf.

| `EnableArm64Rcpc2` | `1` | RCpc2 |
| `EnableArm64Cssc` | `0` | Common Short Sequence Compression (CSSC) |
| `EnableArm64Sve` | `1` | Scalable Vector Extension (SVE) |
| `EnableArm64Sve2` | `1` | SVE2 |
| `EnableArm64Sha3` | `1` | SHA3 |
| `EnableArm64Sm4` | `1` | SM4 |
| `EnableArm64SveAes` | `1` | SVE AES |
| `EnableArm64SveSha3` | `1` | SVE SHA3 |
| `EnableArm64SveSm4` | `1` | SVE SM4 |

A knob defaulting to `0` (for example, `EnableAVX10v2` or `EnableArm64Cssc`) gates an instruction set that's still coming online, so it stays off until you opt in.

## Benchmark to confirm the win

> Vectorization adds complexity, so measure that it pays off before you keep it. Use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet), and use the same environment variables above to compare the scalar, `Vector128`, and `Vector256` implementations in one run. BenchmarkDotNet's disassembly diagnoser can also emit the generated assembly, which is invaluable when tuning high-performance code.

Reviewer: ```suggestion
Vectorization adds complexity, so measure that it pays off before you keep it. Use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet), and use the same environment variables shown previously to compare the scalar, `Vector128`, and `Vector256` implementations in one run. BenchmarkDotNet's disassembly diagnoser can also emit the generated assembly, which is invaluable when tuning high-performance code.
```

## Best practices

- Reach for the existing higher-level APIs first. `Span<T>`, `string`, LINQ, `TensorPrimitives`, and the tensor types already accelerate many common operations for you—don't hand-roll what's already optimized and tested.
- Start with `Vector128<T>`; it's accelerated on the broadest set of hardware, and you don't need `Vector256<T>` to get a correct, portable implementation. Add wider widths and hardware intrinsics only for measured hot paths.
- Check `IsHardwareAccelerated` and `Count` directly instead of caching them; the JIT turns them into constants.
- Always handle the loop remainder, and account for overlapping source and destination buffers when storing.
- Write edge-case tests first, then a scalar solution, then express that scalar logic with the vector APIs.
- Test every code path (including access violations) and benchmark realistic input sizes before you commit to the added complexity.

## See also

> - <xref:System.Runtime.Intrinsics?displayProperty=fullName>

Reviewer: ```suggestion
- <xref:System.Runtime.Intrinsics>
```
